Add first tutorial lessons (#484)
crystal-lang/crystal-book: docs/tutorials/01_basics/06_methods.md

# Methods

To avoid duplication of the same message, instead of using a variable we can
define a method and call it multiple times.

A method definition is indicated by the keyword `def` followed by the method name.
Every expression until the keyword `end` is part of the method body.

```{.crystal .crystal-play}
def say_hello
  puts "Hello Penny!"
end

say_hello
say_hello
say_hello() # syntactically equivalent method call with parenthesis
```

!!! tip
    Method calls are unambiguously indicated by parenthesis after the name, but it can be omitted. It would only be
    necessary for disambiguation, for example if `say_hello` was also a local variable.

## Arguments

What if we want to greet different people, but all in the same manner?
Instead of writing individual messages, we can define a method that allows customization through a parameter.

> A parameter is like a local variable inside the method body. Parameters are declared after the method name in parenthesis.

Reviewer: ```suggestion
A parameter is like a local variable inside the method body. Parameters are declared after the method name in parentheses.
```

**And I think all of the usages in this file should be "parentheses" - plural**

Type restrictions limit the allowed type of an argument. They come after the argument name, separated by a colon:

```{.crystal .crystal-play}
def say_hello(recipient : String)
  puts "Hello #{recipient}!"
end

say_hello "World"
say_hello "Crystal"

# Now this expression doesn't compile:
# say_hello 6
```

Now names cannot be numbers or other data types anymore. This doesn't mean you can't
greet people with a number as a name. The number just needs to be expressed as a string.
Try `say_hello "6"` for example.

## Overloading

Restricting the type of an argument can be used for positional overloading.
When a method has an unrestricted argument like `say_hello(recipient)`, *all* calls to a method `say_hello` go to that method.
But with overloading several methods of the same name can exist with different argument type restrictions. Each call is routed
to the most fitting overload.

```{.crystal .crystal-play}
# This methods greets *recipient*.
def say_hello(recipient : String)
  puts "Hello #{recipient}!"
end

# This method greets *times* times.
def say_hello(times : Int32)
  puts "Hello " * times
end

say_hello "World"
say_hello 3
```

> Overloading is not only defined by type restrictions. The number of arguments as well as named arguments are also

Reviewer: ```suggestion
Overloading isn't defined just by type restrictions. The number of arguments as well as named arguments are also
```
